### DISCUSSION TITLE
How to provide a starting point for BMIBNB in yalmip?

### DISCUSSION THREAD

**Discussion by namrata-kundu** [yalmip/YALMIP#1208] [Possible bug or missing functionality]
I have a non-convex optimization problem and bmibnb takes a very long time to run and times out. Is there a way to provide a custom starting point such that it can converge to the optimal solution faster?

**Comment by johanlofberg** [comment]
https://yalmip.github.io/warmstart

Most likely though, the problem is not to find a good solution (or even the globally optimal one) but proving it is optimal by closing the gap from below.

**Reply by namrata-kundu** [reply]
Thanks so much for sharing the link. However, when I try the warmstart with bmibnb as solver, it errors out stating: 'You have specified an initial point, but the selected solver (MOSEK-SDP) does not support warm-starts through YALMIP'
Sharing a screenshot for your reference:
![Screenshot_20230111_124621](https://user-images.githubusercontent.com/19433397/211759740-50a5b1a7-0577-46e1-9307-c20aff2e36a0.png)

Is there a different non-convex global solver that does support warmstart? Or can I customize sdpsettings in a certain way for bmibnb to take the initial points into consideration?

**Reply by johanlofberg** [reply]
It appears the experimental SDP-based strengthening of a indefinite QP gets confused with that setting.  At the moment, you will have to hack the function addShiftedQP in branch_and_bound and add ops.usex0=0 before the failing line (or set bmibnb.lowerpsdfix to 0 in your options to turn off this presolve strategy)

**Reply by namrata-kundu** [reply]
Hello, thank you, I could avoid the above error using your tip, however, now I face a different error: "Unrecognized function or variable 'relgap'." and on debugging I see, the program isn't entering any of the conditions that set 'relgap' (above the error line) which is why the error. Can you please help? Below are the screenshots for your reference. 
![image](https://user-images.githubusercontent.com/19433397/220467984-d2f0adf7-f7a6-4e3d-91f2-6be33ccb5faa.png)
![image](https://user-images.githubusercontent.com/19433397/220468064-b4bffa5b-e796-4902-9974-765ada1fee1a.png)

**Reply by johanlofberg** [reply]
Switch to the develop branch, I believe it has been fixed there

**Reply by namrata-kundu** [reply]
I switched to the develop branch (https://github.com/yalmip/yalmip/archive/develop.zip), however, I am still getting the same error - the program isn't entering any of the conditions that sets 'relgap'. Attaching screenshot for reference:
![image](https://user-images.githubusercontent.com/19433397/220561489-6fbfc516-8c52-4cf0-acff-251331c6a07b.png)

**Reply by johanlofberg** [reply]
you will have to supply a reproducible example

**Reply by namrata-kundu** [reply]
I realized this only happens when I have both 'bmibnb.lowerpsdfix' set to 0 in options and have ops.usex0=0 set in the function addShiftedQP in bmibnb_branch_and_bound. Removing the 'bmibnb.lowerpsdfix'=0 in options removes the error. However, now that it doesn't throw the error, I see a different issue. 

I know the model has a feasible solution and when I set all the initial points to the expected solutions using warm start and do a check(constraints), all of them pass and I get the solution I am looking for. However, if I even make a minor change in the initial points, eg: changing -5 to -9, for just one of the variables among 120 sdpvar variables, the model says "Infeasible problem (learn to debug) (BMIBNB)" after doing just 1 iteration. I'm attaching a screenshot for reference.
![image](https://user-images.githubusercontent.com/19433397/221382391-dd1939df-66ed-4f49-8e04-be9c8e241142.png)

If this information isn't enough, I can try to create a reproducible example and share it. Any help is highly appreciated.

**Reply by johanlofberg** [reply]
In the screen shot, at the very top I can see you have used check, and the two visible rows shows an infeasible solution

**Reply by johanlofberg** [reply]
you will have to supply a reproducible example, and a claimed true solution which you can assign (using assign) and check feasibility of (using check)

**Reply by namrata-kundu** [reply]
Hello, I've attached the reproducible example herewith.
The file **feasible_reproducible_example.m** has the feasible solution assigned to the variables as warmstart initial points and the check(constraints) pass here. The expected value of the objective is 28.6361.
The file **nonfeasible_reproducible_example.m** is the exact same code with a minor change in the initial starting point (I've only changed a **-5 to -4**) and it leads to an infeasible solution as seen in the prior comment screenshot. (In the earlier comment, the check(constraint) showed infeasible because the initial point was modified)

Please note: This code uses the yalmip **develop** branch and has **ops.usex0=0** set in the function **addShiftedQP** in bmibnb_branch_and_bound.

Can you please kindly help? Thank you.
[reproducible_example.zip](https://github.com/yalmip/YALMIP/files/10915086/reproducible_example.zip)

**Reply by namrata-kundu** [reply]
Hello, just checking in to see if there's any update here. I would highly appreciate if you could please kindly let me know if this is an issue on yalmip's side or if I'm missing something here. It'll help me move forward.
Thanks for all your help.

**Reply by johanlofberg** [reply]
The numerics here is just simply brutal, so not surprised that weird tings happen

For instance, trying to solve the integer-relaxed problem using fmincon, and it shows horrible logs, both cold- and warm-starting

`sol = optimize(Constraints,objective, sdpsettings('solver','fmincon','usex0',0,'relax',1))`

If you turn on display so you see the logs from the linear relaxations too, you will see that the solver constantly is complaining about poor numerics (large bounds)

`sol = optimize(Constraints,objective, sdpsettings('solver','bmibnb','verbose',3))`

As far as I can see in your model, you lack bounds on many of the decision variables, and this will destroy the numerics in a global b&b solver, and in theory typically make the problem unsolvable using a b&b solver as the branching domain is infinite. You will have to fix this before anything can be said or done

**Reply by johanlofberg** [reply]
Also make sure you are running the develop branch, and make sure you pull for changes, as I am fixing various robustness issues popping up on this bad model

**Reply by johanlofberg** [reply]
The model is strange 

This

```epsilon = 0.00005;
for i=1:size(npp,2)-1
    con1= npp(i) <= (npp(i+1)-epsilon);
    Constraints1 = [Constraints1, con1]:'constraints1';
end
```
immediately says there can be at most 1 zero in npp, which renders the very complicated ` nnz(npp) >= 10`  which forces the use of integer variables and a hard MINLP model, effectively redundant

**Reply by johanlofberg** [reply]
BTW, this

```
a = cell(size(new_f,2),1);
b = cell(size(new_f,2),1);
c = cell(size(new_f,2),1);
for i=1:size(new_f,2)
    a{i} = sdpvar(1,1); 
    b{i} = sdpvar(1,1); 
    c{i} = sdpvar(1,1); 
    values = [a{i}; b{i}; c{i}];
    rho = [rho values];
end

```
is a complicated way of doing

```
n = size(new_f,2);
a = sdpvar(1,n)
b = sdpvar(1,n)
c = sdpvar(1,n)
rho = [a;b;c];

```
